### DISCUSSION TITLE
Trigger like solution for DataGridCell using DataTriggerBehavior (Avalonia.Xaml.Interactivity)

### DISCUSSION THREAD

**Discussion by IgorSakovich** [AvaloniaUI/Avalonia#8121] [Q&A] [answered]
Hello! Please tell me how I can implement trigger for each DataGrid cell related to Age property throw ChangePropertyAction ? I have not yet been able to implement this through table styles. Is this the right decision? Thanks!
```
<DataGrid Name="MyDataGrid" 
				  Items="{Binding Persons}" 
				  AutoGenerateColumns="False">
			
				<!--<i:Interaction.Behaviors>
				<ia:DataTriggerBehavior
					Binding="{Binding Age}"
					ComparisonCondition="Equal"
					Value="13">
					<ia:ChangePropertyAction
						TargetObject="MyDataGrid"
						PropertyName="Background"
						Value="Red" />
				</ia:DataTriggerBehavior >
			</i:Interaction.Behaviors>-->

			<DataGrid.Columns>
				<DataGridTextColumn Header="Forename"	Binding ="{Binding FirstName}"/>
				<DataGridTextColumn Header="Surname"	Binding="{Binding LastName}" />
				<DataGridTextColumn Header="Age"		Binding="{Binding Age}"/>
			</DataGrid.Columns>
		</DataGrid>
```

**Comment by maxkatz6** [comment]
I think you can attach a behavior to the control using styles. @wieslawsoltes does it work with stable branch and is there any example of that?

So, in your case you would need to set CellClasses on the right column (let's say "ageColumn").
And with selector "DataGridCell.ageColumn" you can add your behavior as an attached property in the selector.

**Reply by wieslawsoltes** [reply]
> I think you can attach a behavior to the control using styles. @wieslawsoltes does it work with stable branch and is there any example of that?

Example: https://github.com/wieslawsoltes/AvaloniaBehaviors/blob/f235559d91acb9a2b051ef40a4ae1bb9a0752446/samples/DragAndDropSample/Views/MainWindow.axaml#L84-L90

It works with stable branch 0.10.x

**Comment by Kerminator1973** [comment]
Thanks for the useful hints! As I am looking for the same answer I have reconstructed the code that is close to your hints:

``` csharp
<Window.Styles>
	<Style Selector="DataGridCell.statusColumn">
		<Setter Property="FontSize" Value="24"/>
		<Setter Property="(int:Interaction.Behaviors)">
			<int:BehaviorCollectionTemplate>
				<int:BehaviorCollection>
					<ia:DataTriggerBehavior Binding="{Binding Status}" ComparisonCondition="Equal" Value="Rejected">
						<ia:ChangePropertyAction TargetObject="DataGridCell" PropertyName="Background" Value="Yellow" />
					</ia:DataTriggerBehavior>
				</int:BehaviorCollection>
			</int:BehaviorCollectionTemplate>
		</Setter>
	</Style>
</Window.Styles>
...
<DataGrid AutoGenerateColumns="False" Margin="10"
		  Items="{Binding Banknotes}" SelectedItem="{Binding SelectedBanknote}">
	<DataGrid.Columns>
		<DataGridTextColumn Header="{x:Static p:Resources.NoteId}" Binding="{Binding Id}" />
		<DataGridTextColumn Header="{x:Static p:Resources.NoteCurrency}" Binding="{Binding Currency}" />
		<DataGridTextColumn Header="{x:Static p:Resources.NoteDenomination}" Binding="{Binding Denomination}" />
		<DataGridTextColumn Header="{x:Static p:Resources.Status}" Binding="{Binding Status}" CellStyleClasses="statusColumn" />
	</DataGrid.Columns>
</DataGrid>
```

Let me explain my intentions: I want to attach the behaviour for replacing the Background property to every DataGridCell object that has a "statusColumn" class. I use the BehaviorCollection because it attaches a behaviour to each object according to the selector.

The code almost works. The only problem I have is that the binding doesn't use the DataGridCell's DataContext (the element of the ObservableCollection<Banknote> from the ViewModel), but it uses the DataContext of the whole View/Window (ViewModel). Is there any way to specify that I need the DataGridCell's DataContext and not the View's DataContext?

**Reply by timunie** [reply]
> but it uses the DataContext of the whole View/Window (ViewModel

That suprises me really. Is this something you noticed in the editor (Intellisense) or did you check it while running your App? IntelliSense or `CompiledBindings` need `x:DataType` defined on the Style.

**Reply by Kerminator1973** [reply]
Thanks for your instant reaction!

I found it in the code. There is a message in the output window: _[Binding] Error in binding to 'Avalonia.Xaml.Interactions.Core.DataTriggerBehavior'.'Binding': 'Could not find a matching property accessor for 'Status' on 'BvsDesktopLinux.ViewModels.MainWindowViewModel'' (DataTriggerBehavior #69227)_

When using the binding with any property of the  MainWindowViewModel all cells of the «Status» column have the yellow background colour.

So the both checks are claiming about the same thing: the DataContext of the ChangePropertyAction references to the ViewModel data context.

However, the Developer Console (F12) definitely says that DataGridRow and its children (DataGridCells) refer to the element of the ObservableCollection.

The location of my example is [here](https://github.com/Kerminator1973/BvsDesktopLinux). My apologies, I haven’t translate it into English yet.

**Reply by Kerminator1973** [reply]
The default bindings (based on reflection) have been replaced with the compiled binding in my example:

``` csharp
<Window xmlns="https://github.com/avaloniaui" ...
		xmlns:local="using:BvsDesktopLinux.Models"
		x:DataType="vm:MainWindowViewModel"
		x:CompileBindings="True">
	<Window.Styles>
		<Style Selector="DataGridCell.statusColumn">
			<Setter Property="FontSize" Value="24"/>
			<Setter Property="(int:Interaction.Behaviors)">
				<int:BehaviorCollectionTemplate>
					<int:BehaviorCollection>
						<ia:DataTriggerBehavior Binding="{CompiledBinding Status}" x:DataType="local:Banknote" 
												ComparisonCondition="Equal" Value="Rejected">
							<ia:ChangePropertyAction TargetObject="DataGridCell" PropertyName="Background" Value="Yellow" />
						</ia:DataTriggerBehavior>
					</int:BehaviorCollection>
				</int:BehaviorCollectionTemplate>
			</Setter>
		</Style>
	</Window.Styles>
...
<DataGrid AutoGenerateColumns="False" Items="{CompiledBinding Banknotes}"
		  SelectedItem="{Binding SelectedBanknote}">
	<DataGrid.Columns>
		<DataGridTextColumn Header="{x:Static p:Resources.NoteId}" 
							Binding="{CompiledBinding Id}" x:DataType="local:Banknote" />
		<DataGridTextColumn Header="{x:Static p:Resources.NoteCurrency}" 
							Binding="{CompiledBinding Currency}" x:DataType="local:Banknote" />
		<DataGridTextColumn Header="{x:Static p:Resources.NoteDenomination}" 
							Binding="{CompiledBinding Denomination}" x:DataType="local:Banknote" />
		<DataGridTextColumn Header="{x:Static p:Resources.Status}" 
							Binding="{CompiledBinding Status}" x:DataType="local:Banknote" 
							CellStyleClasses="statusColumn" />
	</DataGrid.Columns>
</DataGrid>
```

The DataGrid works well – there are no visible mistakes in its behavior. Unfortunately I get Binding errors when the «DataTriggerBehavior» is uncommented:

``` log
[Binding] Error in binding to 'Avalonia.Xaml.Interactions.Core.DataTriggerBehavior'.'Binding': 'Null value in expression '{empty}' at ''.' (DataTriggerBehavior #23692278)
```

So, my conclusion is: the DataTriggerBahavior uses the ModelView's DataContext, not the DataContext of a collection element.

**Reply by timunie** [reply]
> The location of my example is [here](https://github.com/Kerminator1973/BvsDesktopLinux). My apologies, I haven’t translate it into English yet. 

would be great if the sample could be a minimal sample written in english. It's hard for me to navigate and understand the App in a language that I don't speak. Thx.

**Reply by Kerminator1973** [reply]
Ok. I made a minimal application that contains only DataGrid and DataTriggerBehavior. Please refer to:
https://github.com/Kerminator1973/Avalonia.DataGridExample

I have written some comments about the code and the problem itself. See the "README.md" file.

**Reply by timunie** [reply]
The sample is pretty clear and good 👍 Thank you for your efforts. Unfortunately, I didn't find a solution yet. I feel like this may be an issue we should check in Avalonia.Behaviors repo. /cc @wieslawsoltes 

As a workaround I changed the Style like this using a Converter. Let me know if 

```cs
public class StatusConverter : IValueConverter
{
    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        var status = value as string;

        return status switch
        {
            "Rejected" => Brushes.Red,
            _ => Brushes.Green
        };
    }

    public object? ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        throw new NotImplementedException();
    }
}
```
<Style Selector="DataGridCell" x:DataType="models:Banknote">
    <Setter Property="Background" Value="{Binding Status, Converter={StaticResource StatusConverter}}" />
</Style>
```xml
<Window.Resources>
    <converter:StatusConverter x:Key="StatusConverter" />
</Window.Resources>

<Style Selector="DataGridCell" x:DataType="models:Banknote">
    <Setter Property="Background" Value="{Binding Status, Converter={StaticResource StatusConverter}}" />
</Style>
```

**Reply by Kerminator1973** [reply]
Thanks for pointing out another way to solve the problem! I’ll try to integrate it into the demo application. My goal is to seeif it is possible to not change the background color of the usual rows. I mean the DataGrid uses something like the «AlternationCount» (WPF) by default and this is great. I would like to keep the style of the usual rows unchanged.

**Reply by timunie** [reply]
there is the :nth-child-Selector. Look up style selector docs for info.

**Reply by Kerminator1973** [reply]
Thank you very much for your great advice! I’ve just changed Brushed.Green to null and I got the desired behaviour (see https://github.com/Kerminator1973/Avalonia.DataGridExample).

**Reply by nasterhe** [reply]
> Thank you very much for your great advice! I’ve just changed Brushed.Green to null and I got the desired behaviour (see https://github.com/Kerminator1973/Avalonia.DataGridExample).

@Kerminator1973 nice work,However, when the reject Status is modified, the color does not follow the change.

**Reply by Kerminator1973** [reply]
> @Kerminator1973 nice work,However, when the reject Status is modified, the color does not follow the change.

I'm really sorry to have taken so long to reply, please give me a little more time. I will try to reproduce the problem and resolve it.

**Reply by Kerminator1973** [reply]
> @Kerminator1973 nice work,However, when the reject Status is modified, the color does not follow the change.

Well, I changed the code of the demo application. Please have a look at the latest commit (58ec696c...). After two seconds, the third banknote (Id=3) changes it's status field from "Rejected" to "Jammed". I used an instance of DispatcherTimer to sleep for 2 seconds in the MainWindow class. In the Timer_Tick method, I change the value of the third note to "Jammed". Please have a look at the model - it inherits the INotifyPropertyChanged interface. It allows you to update the DataGrid in the MainWindow. Does it make sense to you?

**Reply by nasterhe** [reply]
> > @Kerminator1973 nice work,However, when the reject Status is modified, the color does not follow the change.
> 
> Well, I changed the code of the demo application. Please have a look at the latest commit (58ec696c...). After two seconds, the third banknote (Id=3) changes it's status field from "Rejected" to "Jammed". I used an instance of DispatcherTimer to sleep for 2 seconds in the MainWindow class. In the Timer_Tick method, I change the value of the third note to "Jammed". Please have a look at the model - it inherits the INotifyPropertyChanged interface. It allows you to update the DataGrid in the MainWindow. Does it make sense to you?

Thanks.I I've found an easier way to do it, please see: #17790

**Comment by timunie** [comment, marked as answer]
Accepted answer in this comment and following, for anyone interested: https://github.com/AvaloniaUI/Avalonia/discussions/8121#discussioncomment-5151274
